### PR TITLE
[Compression] block sparse refactor

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
 - Put block sparse logic to `Scaling`
 - refactor sparsity allocator (using Scaling)

In the future, we hope using `Scaling` could handle different sparse pattern.
The relevant changes about `pruning_target`/`target_name` are for the refactor of `config_list`.

following tasks:
 - rewrite pruning customize doc

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [x] test case
  - [ ] doc

### CODE REVIEW

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 1]
```diff
@@ -275,12 +276,12 @@ def reset_tools(self):
         else:
             self.data_collector.reset()
         if self.metrics_calculator is None:
-            self.metrics_calculator = NormMetricsCalculator(p=self.p, dim=0)
+            self.metrics_calculator = NormMetricsCalculator(p=self.p, scalors=Scaling(kernel_size=[1], kernel_padding_mode='back'))
         if self.sparsity_allocator is None:
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:280]: what is the relation between `metrics_calculator` and `sparsity_allocator`? I thought `metrics_calculator` is a member of sparsity_allocator.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 2]
```diff
@@ -440,12 +441,12 @@ def reset_tools(self):
         else:
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:441]: the function name `reset_tools` is not very good. what is this function used for?

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 3]
```diff
@@ -440,12 +441,12 @@ def reset_tools(self):
...
         if self.metrics_calculator is None:
-            self.metrics_calculator = DistMetricsCalculator(p=2, dim=0)
+            self.metrics_calculator = DistMetricsCalculator(p=2, scalors=Scaling(kernel_size=[1], kernel_padding_mode='back'))
         if self.sparsity_allocator is None:
             if self.mode == 'normal':
-                self.sparsity_allocator = NormalSparsityAllocator(self, dim=0)
+                self.sparsity_allocator = NormalSparsityAllocator(self, Scaling(kernel_size=[1], kernel_padding_mode='back'))
             elif self.mode == 'dependency_aware':
-                self.sparsity_allocator = Conv2dDependencyAwareAllocator(self, 0, self.dummy_input)
+                self.sparsity_allocator = DependencyAwareAllocator(self, self.dummy_input)
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:449]: `DependencyAwareAllocator` does not need `scalors`?

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 4]
```diff
@@ -783,7 +784,7 @@ def _activation_trans(self, output: Tensor) -> Tensor:
         return torch.eq(self._activation(output.detach()), torch.zeros_like(output)).type_as(output)
 
     def _get_metrics_calculator(self) -> MetricsCalculator:
-        return APoZRankMetricsCalculator(dim=1)
+        return APoZRankMetricsCalculator(Scaling(kernel_size=[-1, 1], kernel_padding_mode='back'))
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:787]: seems every time `_get_metrics_calculator` is called, a new calculator is created. is this an expected behavior?

Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:787]: seems it is an expected behavior. then it might be better to use "create" instead of "get", but anyway, not important

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 5]
```diff
@@ -13,8 +14,8 @@
 from torch.nn import Module
 from torch.optim import Optimizer
 
-from nni.algorithms.compression.v2.pytorch.base import Pruner, LayerInfo, Task, TaskResult
-from nni.algorithms.compression.v2.pytorch.utils import OptimizerConstructHelper
+from ...base import Pruner, LayerInfo, Task, TaskResult
+from ...utils import OptimizerConstructHelper, Scaling
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:18]: why change to relative import?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 6]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
-        then the dimension mask will expand to the final mask `[[[1, 1, 1, 1], [0, 0, 0, 0], [1, 1, 1, 1]], [[1, 1, 1, 1], [0, 0, 0, 0], [1, 1, 1, 1]]]`.
-    block_sparse_size
-        This used to describe the block size a metric value represented. By default, None means the block size is ones(len(dim)).
-        Make sure len(dim) == len(block_sparse_size), and the block_sparse_size dimension position is corresponding to dim.
-
-        Example:
-
-        The metric size is (12,), and block_sparse_size=[64], then the mask will expand to (768,) at first before expand with `dim`.
+    scalors
+        Scalor is used to scale the masks' size. It shrinks the mask of the same size as the pruning target to the same size as the metric,
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:297]: how about: It groups a bunch of values in target tensor to calculate one metric value for the group.

what confuses me the most is that what "mask size" represents. the same as target tensor size? the size after shrinking?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 7]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
+        or expands the mask of the same size as the metric to the same size as the pruning target.
+        If you want to use different scalors for different pruning targets in different modules, please use a dict `{module_name: {target_name: scalor}}`.
+        If allocator meets an unspecified module name, it will try to use `scalors['_default'][target_name]` to scale its mask.
+        If allocator meets an unspecified target name, it will try to use `scalors[module_name]['_default']` to scale its mask.
+        Passing in a scalor instead of a `dict` of scalors will be treated as passed in `{'_default': {'_default': scalors}}`.
+        Passing in `None` means no need to scale.
     continuous_mask
-        Inherit the mask already in the wrapper if set True.
+        If set True, the part that has been masked will be masked first.
+        If set False, the part that has been masked may be unmasked due to the increase of its corresponding metric.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:306]: where is the previous mask stored?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 8]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
+    def _get_scalor(self, module_name: str, target_name: str) -> Scaling | None:
+        # Get scalor for the specific target in the specific module. Return None if don't find it.
+        # `module_name` is not used in current nni version, will support different modules using different scalors in the future.
+        if self.scalors:
+            default_module_scalors = self.scalors.get('_default', {})
+            default_target_scalor = default_module_scalors.get(target_name, default_module_scalors.get('_default', None))
+            module_scalors = self.scalors.get(module_name, {})
+            return module_scalors.get(target_name, module_scalors.get('_default', default_target_scalor))
+        else:
+            return None
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:323]: seems this function is implemented twice?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 9]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
+
+    def _expand_mask(self, module_name: str, target_name: str, mask: Tensor) -> Tensor:
+        # Expand the shrinked mask to the pruning target size.
+        scalor = self._get_scalor(module_name=module_name, target_name=target_name)
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:327]: you see you have to use`module_name` and `target_name` to index scalors, and you also use the same index to retrieve the mask from wrapper. it may imply that scalors should be stored in wrapper.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 10]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
+
+    def common_target_masks_generation(self, metrics: Dict[str, Tensor]) -> Dict[str, Dict[str, Tensor]]:
         """
+        Generate masks for metrics-dependent targets.
+
         Parameters
         ----------
         metrics
-            A metric dict. The key is the name of layer, the value is its metric.
+            The format is {module_name: weight_metric}.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:358]: do we need "target_name" here?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 11]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
-            The masked module name.
-        mask
-            The reduced mask with `self.dim` and `self.block_sparse_size`.
-
-        Returns
-        -------
-        Dict[str, Tensor]
-            The key is `weight` or `bias`, value is the final mask.
+        masks
+            The format is {module_name: {target_name: mask}}.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:377]: I noticed "{module_name: {target_name: mask}}" in many places, is it better to define a data structure for it?

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:441]: yes, try to remove it, it is used for resetting the `DataCollector`, `MetricCalculator`, `SparsityCalculator` when resetting the pruner.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 3 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:449]: only channel pruning is meaningful in `DependencyAwareAllocator`, but you are right, the `scalors` should be exposed in the interface.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 5 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:18]: I wish we could move folder `nni/algorithms/compression/v2/pytorch/` to `nni/compression/pytorch/` in `nni v3.0`, so changed the path containing `v2` to relative path.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/metrics_calculator.py [hunk 12]
```diff
@@ -30,47 +33,19 @@ class NormMetricsCalculator(MetricsCalculator):
...
-                So in this case, `dim=0` will set in `__init__`.
-
-                Case 2: Use the output of the conv module as data to calculate the metric.
-                Then the data has size (batch_num, 32, feature_map_size_1, feature_map_size_2).
-                Mention that the dimension 1 of the data is corresponding to the under pruning weight dimension 0.
-                So in this case, `dim=1` will set in `__init__`.
-
-            In both of these two case, the metric of this module has size (32,).
-        p
-            The order of norm. None means Frobenius norm.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/metrics_calculator.py:59]: this docstring is removed, I think it is helpful.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 7 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:306]: previous mask is, e.g., `wrapper.weight_mask`
after all mask's generation done, then update the `wrapper.weight_mask`

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 8 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:323]: yes, the original plan was to put the function into the wrapper after the enhanced pr of the config list, and we use it like `wrapper.metric_calculation_scalor`

I reconsider we discussed in the review meeting, I think `Scaling` should be a class, but maybe it should not in wrapper, I will give some proposals in next meeting.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 9 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:327]: yes, this is our final goal, but not in this pr

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 13]
```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
...
+        # Set the already masked part in the metric to the minimum value.
+        target_name = 'weight'
+        for module_name, target_mask in new_masks.items():
+            wrapper = self.pruner.get_modules_wrapper()[module_name]
+            old_target_mask = getattr(wrapper, f'{target_name}_mask', None)
+            if old_target_mask is not None:
+                new_masks[module_name][target_name] = torch.min(target_mask[target_name], old_target_mask)
+        return new_masks
+
+    def common_target_masks_generation(self, metrics: Dict[str, Tensor]) -> Dict[str, Dict[str, Tensor]]:
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:351]: `common`: means generate masks by data->metric->mask.
`special`: now we have `bias_mask = (0 if output channel all masked else 1)`

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 10 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:358]: we need, but not this pr

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 11 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:377]: good suggestion, will add one, but maybe `{module_name: {target_name: mask}}` maybe is more friendly for review.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 5 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:18]: got it

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 14]
```diff
@@ -13,12 +14,24 @@
...
 
-from nni.algorithms.compression.v2.pytorch.base import Pruner, LayerInfo, Task, TaskResult
-from nni.algorithms.compression.v2.pytorch.utils import OptimizerConstructHelper
+from ...base import Pruner, LayerInfo, Task, TaskResult
+from ...utils import OptimizerConstructHelper, Scaling
 
 _logger = logging.getLogger(__name__)
 
 
+def _get_scalor(scalors: Dict[str, Dict[str, Scaling]] | None, module_name: str, target_name: str) -> Scaling | None:
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:23]: scalor or scalar or scaler?

File: nni/algorithms/compression/v2/pytorch/utils/common.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/common.py:1]: common in utils doesn't make much sense. Looks like "attr.py"

File: test/algo/compression/v2/test_pruning_tools_torch.py [hunk 16]
```diff
@@ -112,7 +113,7 @@ def collect_taylor(grad):
 
     def test_metrics_calculator(self):
         # Test NormMetricsCalculator
-        metrics_calculator = NormMetricsCalculator(dim=0, p=2)
+        metrics_calculator = NormMetricsCalculator(p=2, scalors=Scaling(kernel_size=[1], kernel_padding_mode='back'))
```
Comment (ultmaster) [inline test/algo/compression/v2/test_pruning_tools_torch.py:116]: Is this public API change?

**Review by ultmaster — APPROVED** [review]

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 14 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py]: oh...it's scaler, fixed

File: nni/algorithms/compression/v2/pytorch/utils/common.py (hunk 15 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/common.py]: good suggestion, rename it

File: test/algo/compression/v2/test_pruning_tools_torch.py (hunk 16 above)
Comment (J-shang) [reply test/algo/compression/v2/test_pruning_tools_torch.py]: it is internal API

**Review by QuanluZhang — APPROVED** [review]
